File: src/bioinformatika/SaisStep.java
package bioinformatika;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.TreeMap;
import java.util.Map.Entry;

public class SaisStep {

	private String[] S;
	private boolean[] t;
	private ArrayList<Integer> lmsPointersArray;
	private TreeMap<String, ArrayList<Integer>> bucketHash;
	private Object[] keySet;
	private HashMap<String, Integer> bucketPointers;
	private String[] arrayWithNewNames;
	private ArrayList<Integer> SA;
	
	public ArrayList<Integer> getSA() {
		return SA;
	}

	public void setSA(ArrayList<Integer> sA) {
		SA = sA;
	}

	public SaisStep(String[] s, boolean[] t,
			ArrayList<Integer> lmsPointersArray,
			TreeMap<String, ArrayList<Integer>> bucketHash, Object[] keySet,
			HashMap<String, Integer> bucketPointers, String[] arrayWithNewNames) {
		super();
		this.S = s;
		this.t = t;
		this.lmsPointersArray = lmsPointersArray;
		this.bucketHash = bucketHash;
		this.keySet = keySet;
		this.bucketPointers = bucketPointers;
		this.arrayWithNewNames = arrayWithNewNames;
	}

	public SaisStep() {
		// TODO Auto-generated constructor stub
	}

	public String[] getArrayWithNewNames() {
		return arrayWithNewNames;
	}

	public void setArrayWithNewNames(String[] arrayWithNewNames) {
		this.arrayWithNewNames = arrayWithNewNames;
	}

	public String[] getS() {
		return S;
	}

	public void setS(String[] s) {
		S = s;
	}

	public boolean[] getT() {
		return t;
	}

	public void setT(boolean[] t) {
		this.t = t;
	}

	public ArrayList<Integer> getLmsPointersArray() {
		return lmsPointersArray;
	}

	public void setLmsPointersArray(ArrayList<Integer> lmsPointersArray) {
		this.lmsPointersArray = lmsPointersArray;
	}

	public TreeMap<String, ArrayList<Integer>> getBucketHash() {
		return bucketHash;
	}

	public void setBucketHash(TreeMap<String, ArrayList<Integer>> bucketHash) {
		this.bucketHash = bucketHash;
	}

	public Object[] getKeySet() {
		return keySet;
	}

	public void setKeySet(Object[] keySet) {
		this.keySet = keySet;
	}

	public HashMap<String, Integer> getBucketPointers() {
		return bucketPointers;
	}

	public void setBucketPointers(HashMap<String, Integer> bucketPointers) {
		this.bucketPointers = bucketPointers;
	}


	public String[] getArrayWithNewNamesForEachLMSSubstring(ArrayList<Integer> lmsPointersArray,
			TreeMap<String, ArrayList<Integer>> bucketHash, String[] S, boolean[] t) {
	
		printArray(S, "Entered getArrayWithNewNames...");
		
		String[] S1 = new String[lmsPointersArray.size()];
 		ArrayList<Integer> SA = new ArrayList<>();
 	
		for (Entry<String, ArrayList<Integer>> entry : bucketHash.entrySet()) {
			for (Integer integerInBucket : entry.getValue()) {
				SA.add(integerInBucket);
			}
		}
	
		int newName = 0;
		System.out.println("lmsPointers: " + lmsPointersArray);
		
		int lastKnownIndexOfLMSSubstringInSA = 0;
		for (int i = 0; i < SA.size(); i++) {
			if (lmsPointersArray.contains(SA.get(i))) {
				lastKnownIndexOfLMSSubstringInSA = i;
				break;
			}
		}
		S1[lmsPointersArray.indexOf(SA.get(lastKnownIndexOfLMSSubstringInSA))] = Integer.toString(newName);
		System.out.println("na mjesto " + lastKnownIndexOfLMSSubstringInSA + " stavljamo " + newName);
		
		for (int i = lastKnownIndexOfLMSSubstringInSA + 1; i < SA.size(); i++) {
			if (lmsPointersArray.contains(SA.get(i)) == true) {
				
				if(compareLMSSubstrings(SA.get(i), SA.get(lastKnownIndexOfLMSSubstringInSA), S, t) == true) {
					S1[lmsPointersArray.indexOf(SA.get(i))] = Integer.toString(newName);
				} else {
					newName++;
					S1[lmsPointersArray.indexOf(SA.get(i))] = Integer.toString(newName);
				}
				lastKnownIndexOfLMSSubstringInSA = i;
			} 
		}
		System.out.println("AHHHHHHHHHHHAAAAAAAAAAAAA");
		System.out.println("ovo je SA: " + SA);
		this.printArray(S1, "UNUTAR STVARANJA ARRAYA");
		
		this.SA = SA;
		return S1;
	
	}


	public void doThirdIteration(String[] S, boolean[] t,
			TreeMap<String, ArrayList<Integer>> bucketHash,
			HashMap<String, Integer> bucketPointers, Object[] keySet) {
		
		// Iterate from the end to the beginning
		for (int i = keySet.length - 1; i >= 0; i--) {
			for (int j = bucketHash.get(keySet[i]).size() - 1; j >= 0; j--) {
				int integerInBucket = bucketHash.get(keySet[i]).get(j);
				if (integerInBucket > 0 && t[integerInBucket - 1] == false) {
					bucketHash.get(S[integerInBucket - 1]).set(bucketPointers.get(S[integerInBucket - 1]), integerInBucket - 1);
					bucketPointers.put(S[integerInBucket - 1],
							bucketPointers.get(S[integerInBucket - 1]) - 1);
				}
			}
		}
		
	}


	public void doSecondIteration(String[] S, boolean[] t, TreeMap<String, ArrayList<Integer>> bucketHash, HashMap<String, Integer> bucketPointers) {
		// Iterate over all buckets
		for (Entry<String, ArrayList<Integer>> entry : bucketHash.entrySet()) {
			for (Integer integerInBucket : entry.getValue()) {
				if (integerInBucket > 0 && t[integerInBucket - 1] == true) {
					bucketHash.get(S[integerInBucket - 1]).set(
							bucketPointers.get(S[integerInBucket - 1]),
							integerInBucket - 1);
					bucketPointers.put(S[integerInBucket - 1],
							bucketPointers.get(S[integerInBucket - 1]) + 1);
				}
			}
		}

	}


	/*
	 * Create SA buckets and insert -1 for each value
	 * */
	public TreeMap<String, ArrayList<Integer>> getBucketHashFromSAndInitializeToMinusOne(String[] S) {

		TreeMap<String, ArrayList<Integer>> bucketHash = new TreeMap<>();
		for (int i = 0; i < S.length; i++) {
			if (bucketHash.containsKey(S[i])) {
				bucketHash.get(S[i]).add(-1);
			} else {
				ArrayList<Integer> values = new ArrayList<>();
				values.add(-1);
				bucketHash.put(S[i], values);
			}
		}
		return bucketHash;
	}
	
	public void setSAToMinusOne(TreeMap<String, ArrayList<Integer>> bucketHash) {
		
 		for (Entry<String, ArrayList<Integer>> entry : bucketHash.entrySet()) {
 	        for (int i = 0; i < entry.getValue().size(); i++) {
 	        	entry.getValue().set(i, -1);
 	        }   
 	    }
		
	}

	public String joinArray(String[] aArr) {
	    StringBuilder sbStr = new StringBuilder();
	    for (int i = 0, il = aArr.length; i < il; i++) {
	        sbStr.append(aArr[i]);
	    }
	    return sbStr.toString();
	}

	public String joinArray(int[] aArr) {
	    StringBuilder sbStr = new StringBuilder();
	    for (int i = 0, il = aArr.length; i < il; i++) {
	        sbStr.append(aArr[i]);
	    }
	    return sbStr.toString();
	}
	
	public boolean compareLMSSubstrings(int firstLMS, int secondLMS, String[] S, boolean[] t) {
		boolean isFirstIteration = true;
		while (S[firstLMS].equals(S[secondLMS])) {

			// Do only once
			if (isFirstIteration == true) {
				if (t[firstLMS] == false && t[secondLMS] == false) {
					firstLMS++;
					secondLMS++;
					// Check if we're comparing last lms strings
					if (firstLMS == secondLMS && firstLMS >= S.length) {
						return true;
					}
					isFirstIteration = false;
					continue;
				}
			}
			
			
			if (t[firstLMS] == false && t[secondLMS] == false) {
				return true;
			} else if (t[firstLMS] == true && t[secondLMS] == true) {
				firstLMS++;
				secondLMS++;
			} else {
				return false;
			}
		
		}
		
		return false;
		
	}
	
	public void setAllPointersInBucketsToBeginning(Object[] keySet, HashMap<String, Integer> bucketPointers) {
		for (int i = 0; i < keySet.length; i++) {
			bucketPointers.put((String) keySet[i], 0);
		}
	}
	
	public void setAllPointersInBucketsToEnd(Object[] keySet, HashMap<String, Integer> bucketPointers, TreeMap<String, ArrayList<Integer>> bucketHash) {
		for (int i = 0; i < keySet.length; i++) {
			bucketPointers.put((String) keySet[i], bucketHash.get((String) keySet[i]).size() - 1);
		}
	}
	
	/*	Returns array of booleans that tells us the types of characters
	 * 	L = true
	 * 	S = false
	 * */
	public boolean[] getTFromS(String[] S) {
		
		boolean[] t = new boolean[S.length];
		t[S.length - 1] = false;
		t[S.length - 2] = true;
		
		for (int i = S.length - 3; i > 0; i--) {
			int comparation = S[i].compareTo(S[i + 1]);
			
			if (comparation < 0) {
				t[i] = false;
			} else if (comparation > 0) {
				t[i] = true;
			} else {
				t[i] = t[i + 1];
			}
		}
		
		return t;
	}

	/*
	 * 	Set array from "i" to "j" to booleanValue
	 * 
	 * */
	private void setArray(boolean[] t, int i, int j, boolean booleanValue) {
		for (; i <= j; i++) {
			t[i] = booleanValue;
		}
	}

	/*
	 * Returns array holding pointers to LMS-substrings
	 * String[] S is the input
	 * boolean[] t keeps track of S-types and L-types
	 * S type = false
	 * L type = true
	 * */
	public ArrayList<Integer> getLMSPointersArray(String[] S, boolean[] t) {
		ArrayList<Integer> lmsPointers = new ArrayList<>();
		for (int i = 1; i < S.length; i++) {
			if (t[i - 1] == true && t[i] == false) {
				lmsPointers.add(i);
			}
		}
		return lmsPointers;
	}

	public static void printArray(String[] arr, String message) {
		System.out.println(message);
		for (int i = 0; i < arr.length; i++) {
			System.out.print(arr[i]);
		}
		System.out.println("");
	}
	
}

File: saisStep.py
class SaisStep:
        def __init__(self, S = [], t = [], lmsPointersArray = [],
                     bucketHash = dict(), keySet = [], bucketPointers = dict(),
                     arrayWithNewNames = [], SA = [], lmsPointersHash = dict()):
                self.S = S
                self.t = t
                self.lmsPointersArray = lmsPointersArray
                self.bucketHash = bucketHash
                self.keySet = keySet
                self.bucketPointers = bucketPointers
                self.arrayWithNewNames = arrayWithNewNames
                self.SA = SA
                self.lmsPointersHash = lmsPointersHash
                

        def getArrayWithNewNames(self):
                return self.arrayWithNewNames
        

        def setArrayWithNewNames(self, arrayWithNewNames):
                self.arrayWithNewNames = arrayWithNewNames
        

        def  getS(self):
                return self.S
        

        def  setS(self, s):
                self.S = s
        

        def  getT(self):
                return self.t
        

        def  setT(self, t):
                self.t = t
        

        def  getLmsPointersArray(self):
                return self.lmsPointersArray
        

        def  setLmsPointersArray(self, lmsPointersArray):
                self.lmsPointersArray = lmsPointersArray
        

        def  getBucketHash(self):
                return self.bucketHash

        def  setBucketHash(self, bucketHash):
                self.bucketHash = bucketHash
        

        def getKeySet(self):
                return self.keySet
        

        def setKeySet(self, keySet):
                self.keySet = keySet
        

        def getBucketPointers(self):
                return self.bucketPointers
        

        def  setBucketPointers(self, bucketPointers):
                self.bucketPointers = bucketPointers
        
        def getArrayWithNewNamesForEachLMSSubstring(self, bucketHash, S, t):
                print ("Getting new names...")
        
                S1 = []
                SA = []

                for i in range(0, len(self.lmsPointersHash)):
                        S1.append("")
                
                for key in sorted(bucketHash.keys()):
                        for InBucket in bucketHash[key]:
                                SA.append(InBucket)
                        
                newName = 0
                lastKnownIndexOfLMSSubInSA = 0
                for i in range(0, len(SA)):
                        if str(SA[i]) in self.lmsPointersHash.keys():
                                lastKnownIndexOfLMSSubInSA = i
                                break
                        
                S1[self.lmsPointersHash[str(SA[lastKnownIndexOfLMSSubInSA])]] = str(newName)
                
                for i in range(lastKnownIndexOfLMSSubInSA + 1, len(SA)):
                        if str(SA[i]) in self.lmsPointersHash:
                                if self.compareLMSSubstrings(SA[i], SA[lastKnownIndexOfLMSSubInSA], S, t) == True:
                                        S1[self.lmsPointersHash[str(SA[i])]] = str(newName)
                                else:
                                        newName += 1
                                        S1[self.lmsPointersHash[str(SA[i])]] = str(newName)
                                lastKnownIndexOfLMSSubInSA = i
                         
        
                print ("Done finding new names")
                self.SA = SA
                return S1

        def doThirdIteration(self, S, t,
                        bucketHash,
                        bucketPointers, keySet):
                
                # Iterate from the end to the beginning
                for ii in range(0, len(keySet)):
                        i = len(keySet) - 1 - ii
                   
                        for jj in range(0, len(bucketHash[str(keySet[i])])):
                                j = len(bucketHash[str(keySet[i])]) - 1 - jj
                                InBucket = bucketHash[str(keySet[i])][j]
                                if InBucket > 0 and t[InBucket - 1] == False:
                                        bucketHash[str(S[InBucket - 1])][bucketPointers[str(S[InBucket - 1])]] = InBucket - 1
                                        bucketPointers[str(S[InBucket - 1])] = bucketPointers[str(S[InBucket - 1])] - 1
                
        def doSecondIteration(self, S, t, bucketHash, bucketPointers):
                # Iterate over all buckets
                
                for key in sorted(bucketHash.keys()):
                        for InBucket in bucketHash[key]:
                                if InBucket > 0 and t[InBucket - 1] == True:
                                        bucketHash[str(S[InBucket - 1])][bucketPointers[str(S[InBucket - 1])]] = InBucket - 1
                                        bucketPointers[str(S[InBucket - 1])] = bucketPointers[str(S[InBucket - 1])] + 1
               
        #
        # Create SA buckets and insert -1 for each value
        #
        def getBucketHashFromSAndInitializeToMinusOne(self, S):

                bucketHash = dict()
                for i in range(0, len(S)):
                        
                        if str(S[i]) in bucketHash.keys():
                                bucketHash[str(S[i])].append(-1)
                        else:
                                values = []
                                values.append(-1)
                                bucketHash[str(S[i])] = values
        
                return bucketHash

        def setSAToMinusOne(self, bucketHash):
                for key in sorted(bucketHash.keys()):
                        for i in range(0, len(bucketHash[key])):
                                bucketHash[key][i] = -1
                   
            
        def  joinArray(self, aArr):
            return ''.join(str(x) for x in aArr)
        
        
        def  compareLMSSubstrings(self, firstLMS, secondLMS, S, t):
                isFirstIteration = True
                while S[firstLMS] == S[secondLMS]:

                        # Do only once
                        if isFirstIteration == True:
                                if t[firstLMS] == False and t[secondLMS] == False:
                                        firstLMS += 1
                                        secondLMS += 1
                                        # Check if we're comparing last lms s
                                        if firstLMS == secondLMS and firstLMS >= len(S):
                                                return True
                                        isFirstIteration = False
                                        continue
                                
                        
                        
                        if t[firstLMS] == False and t[secondLMS] == False:
                                return True
                        elif t[firstLMS] == True and t[secondLMS] == True:
                                firstLMS += 1
                                secondLMS += 1
                        else:
                                return False
                        
                
                return False
        
        
        def setAllPointersInBucketsToBeginning(self, keySet, bucketPointers):
               for i in range(0, len(keySet)):
                        bucketPointers[str(keySet[i])] = 0
                
        
        
        def setAllPointersInBucketsToEnd(self, keySet, bucketPointers, bucketHash):  
                for i in range(0, len(keySet)):
                        bucketPointers[str(keySet[i])] = len(bucketHash[str(keySet[i])]) - 1
              
        
        
        #      Returns array of s that tells us the types of characters
        #      L = True
        #      S = False
        # 
        def getTFromS(self, S):
               
                t = []
                for i in range(0, len(S)):
                        t.append(False)
                t[len(S) - 1] = False
                t[len(S) - 2] = True

                for ii in range(1, len(S) - 2):
                        i = len(S) - 2 - ii
                        if S[i] < S[i + 1]:
                                t[i] = False
                        elif S[i] > S[i + 1]:
                                t[i] = True
                        else:
                                t[i] = t[i + 1]
                        
                
                return t
        

        #
        #      Set array from "i" to "j" to Value
        # 
        #
        def setArray(self, t, i, j,  Value):
                while i <= j:
                        t[i] = Value
                        i += 1
             
        #
        # Returns array holding pointers to LMS-subs
        # S is the input
        # [] t keeps track of S-types and L-types
        # S type = False
        # L type = True
        # 
        def  getLMSPointersHash(self, S, t):
                lmsPointersHash = dict()
                lmsPointers = []
                for i in range(1, len(S)):
                        if t[i - 1] == True and t[i] == False:
                                lmsPointers.append(i)
                                lmsPointersHash[str(i)] = len(lmsPointers) - 1
                                
                self.lmsPointersHash = lmsPointersHash
                self.lmsPointersArray = lmsPointers
                return lmsPointers
        

        def  getLmsPointersHash(self):
                return self.lmsPointersHash
        

        def setLmsPointersHash(self, lmsPointersHash):
                self.lmsPointersHash = lmsPointersHash
        

        def  printArray(self, arr,  message):
                print (message)
                print (arr)
        
